File: project_euler/Problem040/Problem040.java
package project_euler;

import java.util.ArrayList;
import java.util.List;

public class Problem040 {
    public void solution() {
        int target = 1000000;
        List<Integer> borders = new ArrayList<>();
        int index = 0;
        while (true) {
            int value = ((int) Math.pow(10, index) * (9 * index - 1) + 1) / 9;
            borders.add(value);
            if (value > target) {
                break;
            }
            ++index;
        }

        int result = 1;
        while (0 != target) {
            result *= getChampernowne(target, borders);
            target /= 10;
        }
        System.out.println(result);
    }

    private int getChampernowne(int value, List<Integer> borders) {
        int index = getIndex(value, borders) - 1;
        value -= borders.get(index) + 1;
        int digit = (int) Math.pow(10, index) + value / (index + 1);
        for (int i = value % (index + 1); i < index; ++i) {
            digit /= 10;
        }
        return digit % 10;
    }

    private int getIndex(int value, List<Integer> borders) {
        int start = 0;
        int end = borders.size() - 1;
        while (start <= end) {
            int current = start + ((end - start) >> 1);
            if (borders.get(current) <= value) {
                start = current + 1;
            } else {
                end = current - 1;
            }
        }
        return start;
    }
}

File: project_euler/Problem040/Problem040.py
from typing import List

class Problem040:
    def solution(self) -> None:
        target, borders, index = 1000000, [], 0
        while True:
            value = (10 ** index * (9 * index - 1) + 1) // 9
            borders.append(value)
            if value > target:
                break
            index += 1

        result = 1
        while target:
            result *= self.__get_champernowne(target, borders)
            target //= 10
        print(result)

    def __get_champernowne(self, value: int, borders: List[int]) -> int:
        index = self.__get_index(value, borders) - 1
        value -= borders[index] + 1
        digit = 10 ** index + value // (index + 1)
        for i in range(value % (index + 1), index):
            digit //= 10
        return digit % 10

    def __get_index(self, value: int, borders: List[int]) -> int:
        start, end = 0, len(borders) - 1
        while start <= end:
            current = start + ((end - start) >> 1)
            if borders[current] <= value:
                start = current + 1
            else:
                end = current - 1
        return start
